### PR TITLE
MDS Date cell

### PR DESCRIPTION
**Pull request by juanca**
https://www.pivotaltracker.com/story/show/180679135

## Note for reviewer(s)

I ended up upgrading a bunch of testing libraries to have a better label on the calendar widget. Unfortunately, it didn't work because there is a [bug with something inside testing-library](https://github.com/testing-library/dom-testing-library/issues/1103). For now, I changed the label on the buttons to be generic. 

## PR upkeep checklist

- [x] Change log entry
- [x] Label(s)
- [x] Assignee(s)
- [x] Deployment URL: https://mavenlink.github.io/design-system/mds-date-cell/
- [x] (When ready for review) Reviewer(s)
- [x] Green Bigmaven CI check: https://github.com/mavenlink/mavenlink/pull/27963
- [ ] DevQA on Bigmaven staging (e.g. does the work need to be imported in the internal design system?)

### CODE REVIEW

File: src/components/autocompleter/autocompleter.test.jsx [hunk 1]
```diff
@@ -45,7 +45,7 @@ describe('src/components/autocompleter/autocompleter', () => {
   describe('validationMessage', () => {
     it('sets the input to be invalid', () => {
       render(<Autocompleter {...requiredProps} validationMessage="not valid" />);
-      expect(screen.getByLabelText('Test label')).toHaveDescription('not valid');
+      expect(screen.getByLabelText('Test label')).toHaveAccessibleDescription('not valid');
```
Comment (juanca) [inline src/components/autocompleter/autocompleter.test.jsx:48]: This is needed for the latest testing-library libs. I did a global find and replace.

File: src/components/cell-control/autocompleter.md [hunk 2]
```diff
@@ -6,22 +6,22 @@ import { Table, TableHeader, TableHeaderCell, TableBody, TableRow, TableCell } f
 
 const ids = {
   th: {
-    autocompleter: uuid.v4(),
+    example: uuid.v4(),
```
Comment (juanca) [inline src/components/cell-control/autocompleter.md:9]: I fixed a bunch of bad IDs in my recent examples. I made it more copy-pastable for now. Ideally, we should have code that _verifies_ the existence of labels. But that's another problem to solve in the future.

File: src/components/custom-field-input-date/__snapshots__/custom-field-input-date.test.jsx.snap [hunk 3]
```diff
@@ -4,58 +4,54 @@ exports[`src/components/custom-field-input-date/custom-field-input-date classNam
 <body>
   <div>
     <div
-      class="container"
```
Comment (juanca) [inline src/components/custom-field-input-date/__snapshots__/custom-field-input-date.test.jsx.snap:7]: In order to extract a "date control" to be shared between the field and cell components, I "simplified" the structure. Just moved some `<div>`s inside and moved the `<FormControl>` to the outside. We lost a little bit of padding but I think that's okay given the easy code re-use.

File: src/components/control/date.jsx [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    event.preventDefault();
+    setActive(true);
+    setEditing(true);
+    setExpanded(true);
+  }
+
+  function onInputKeyDown(event) {
+    setActive(true);
+
+    if (props.readOnly) return;
```
Comment (Bloodyaugust) [inline src/components/control/date.jsx:86]: Should we be able to `setActive` if the input is `readOnly`?

File: src/components/control/date.jsx [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+      target: ref.current,
+      detail: {
+        validationMessage,
+      },
+    });
+  }, [validationMessage]);
+
+  useLayoutEffect(() => {
+    refs.input.current.value = editing ? toFullDateFormat(value) : toDateStringFormat(value);
+  }, [value]);
```
Comment (Bloodyaugust) [inline src/components/control/date.jsx:141]: Should this dep array also include `editing`? 🤔

**Review by Bloodyaugust — APPROVED** [review]
LGTM! There are some non-blocking questions and potential refactors.

File: package.json [hunk 6]
```diff
@@ -34,10 +34,10 @@
     "@babel/plugin-proposal-class-properties": "^7.2.3",
     "@babel/preset-env": "^7.2.3",
     "@babel/preset-react": "^7.0.0",
-    "@testing-library/jest-dom": "^5.11.1",
-    "@testing-library/react": "^10.4.7",
-    "@testing-library/react-hooks": "^3.4.1",
-    "@testing-library/user-event": "^13.1.8",
+    "@testing-library/jest-dom": "^5",
+    "@testing-library/react": "^12",
```
[the remark this answers is not in the document]
Comment (juanca) [reply package.json:38]: Hmmmm.... I don't like the extra work. Theoretically, both suites of testes are independent of each other. I would prefer less work 😅  and or less friction for updating libs.

File: src/components/calendar/calendar.css [hunk 7]
```diff
@@ -13,7 +13,6 @@
   padding-left: var(--spacing-medium);
   padding-right: var(--spacing-medium);
   width: calc((var(--spacing-x-large) * 7) + calc(var(--spacing-medium) * 2));
-  height: 293px;
```
[the remark this answers is not in the document]
Comment (juanca) [reply src/components/calendar/calendar.css:16]: I'm actually confused why this was hard-coded. It looks 👨‍🍳 👌  without it. 🤔

File: src/components/control/date.jsx [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+import calendarSvg from '../../svgs/calendar.svg';
+import Calendar from '../calendar/calendar.jsx';
+import IconButton from '../icon-button/icon-button.jsx';
+import Icons from './icons.jsx';
+import useDropdownClose from '../../hooks/use-dropdown-close.js';
+import styles from './date.css';
+import useForwardedRef from '../../hooks/use-forwarded-ref.js';
+
+function toDateStringFormat(date) {
+  return date ? date.toLocaleDateString(undefined, { month: 'short', year: 'numeric', day: 'numeric' }) : '';
```
[the remark this answers is not in the document]
Comment (juanca) [reply src/components/control/date.jsx:19]: That probably makes sense but I don't know of a real use-case to let it be a prop. I would rather keep it good enough. Let me look into it.

File: src/components/control/date.jsx [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+function toFullDateFormat(date) {
+  return date ? date.toISOString().slice(0, 10) : undefined;
```
[the remark this answers is not in the document]
Comment (juanca) [reply src/components/control/date.jsx:23]: Nope! Pre-existing code and I think it is working?

File: src/components/control/date.jsx (hunk 4 above)
Comment (juanca) [reply src/components/control/date.jsx:86]: 🤔 I'm not sure how `active` is used here but a read-only input can still be focusable which I think implies it is active?

File: src/components/control/date.jsx (hunk 5 above)
Comment (juanca) [reply src/components/control/date.jsx:141]: I don't think it matters here:

- If the value changes then we re-format and update the DOM reference
- If editing changes, we are already re-formatting the in-memory value and passing it down as a `value` or `defaultValue` prop IIRC

File: src/components/control/date.jsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (juanca) [reply src/components/control/date.jsx:19]: `undefined` results in browser defaults for now. I think that's pretty good until we store the user's preference in our data model.

File: package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (Bloodyaugust) [reply package.json:38]: Fair point! They are pretty well separated. And I'm all for less work! 😉

File: src/components/calendar/calendar.css (hunk 7 above)
[the remark this answers is not in the document]
Comment (Bloodyaugust) [reply src/components/calendar/calendar.css:16]: I bet it was some edge-case that we'll run into again now that it's gone. 😉

File: src/components/control/date.jsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (Bloodyaugust) [reply src/components/control/date.jsx:19]: Ah, I was under the impression that either project or account did store that, but maybe it's just timezone? 🤔

File: src/components/control/date.jsx (hunk 4 above)
Comment (Bloodyaugust) [reply src/components/control/date.jsx:86]: Fair enough. We can always tweak it later if needed.

File: src/components/control/date.jsx (hunk 5 above)
Comment (Bloodyaugust) [reply src/components/control/date.jsx:141]: Cool, sounds good!

**Pull request merged**